instrumentation/excon/lib/opentelemetry/instrumentation/excon/middlewares/tracer_middleware.rb

```diff
@@ -10,25 +10,16 @@ module Excon
       module Middlewares
         # Excon middleware for instrumentation
         class TracerMiddleware < ::Excon::Middleware::Base
-          HTTP_METHODS_SYMBOL_TO_STRING = {
-            connect: 'CONNECT',
-            delete: 'DELETE',
-            get: 'GET',
-            head: 'HEAD',
-            options: 'OPTIONS',
-            patch: 'PATCH',
-            post: 'POST',
-            put: 'PUT',
-            trace: 'TRACE'
-          }.freeze
+          HTTP_METHODS_TO_UPPERCASE = Hash.new { |hash, key| hash[key] = key.to_s.upcase }
```

Comment:
We'd have to add...
```ruby
  accum[upcased_method] = upcased_method
```

But I wonder whether it's worth the extra code. Would it suffice to make these constants private?

Reply:
Could you help me understand how making them private make the hashes thread safe?

Reply:
Making them private protects them from external modification (assuming sensible code).

As for thread safety, I'm venturing into water that is probably deeper than I can swim here, but I'll try.

On MRI, the Concurrent Ruby library [declares that `Hash` is thread-safe](https://github.com/ruby-concurrency/concurrent-ruby/blob/v1.1.9/lib/concurrent-ruby/concurrent/hash.rb#L17-L21).

On JRuby, parallel threads may try to read the value and add it to the hash if it doesn't already exist. Since the default proc always yields the same value, it's safe if threads overwrite each other. It's even safe if two threads add the same value twice. However, I don't believe this could happen in the [`RubyHash`](https://github.com/jruby/jruby/blob/9.3.3.0/core/src/main/java/org/jruby/RubyHash.java) implementation, which follows set-then-publish and ensures writes to the underlying `table` are atomic. Even when the `table` has to be resized, it's [a `synchronized` operation](https://github.com/jruby/jruby/blob/9.3.3.0/core/src/main/java/org/jruby/RubyHash.java#L468).

I don't know much about TruffleRuby.

_Having said that_, I realise that the safety of the operation is not immediately visible, so I'm now in favour of using a frozen hash as you suggest! :)